### ISSUE TITLE
React 18 doesn't wait for long running suspended promises

### ISSUE THREAD

**Issue by blittle** [facebook/react#22791]
At Shopify with Hydrogen, we are encountering a problem with long running requests and React Server Components. If a component within a suspense boundary makes a request to blackhole.webpagetest.org, `pipeToNodeWritable` doesn't stay open while the promise to blackhole is suspended.

Steps:

1. Request from the client
2. Server starts `pipeToNodeWritable`
3. One component in the tree makes a request to blackhole, which won't ever resolve (node-fetch eventually times out).
4. React almost immediately calls `onCompleteAll`, even though the underlying promise has yet to resolve.
5. `onError` is never called either
6. This error is in the server console: `Warning: There was still abortable task at the root when we closed. This is a bug in React.`

Eventually the error from the request does timeout, and it shows up in the console, but it's too late to stream anything to the client because the response has been closed. But even catching that error is hard/difficult, because. it is just dumped to the console, and doesn't actually go through React's `pipeToNodeWritable` `onError` handler.

See the example here: https://stackblitz.com/edit/hydrogen-v3w1xm?file=src%2Fpages%2FIndex.server.jsx

Note in Index.server.jsx:73 it's fetching data to blackhole. Note the warnings in the console:

> Warning: There was still abortable task at the root when we closed. This is a bug in React.

**Comment by artem-malko** [comment]
Hi) I'll try to help you with your problem.

First of, you have to migrate from `pipeToNodeWritable` to `renderToPipeableStream` according to this [link](https://github.com/reactwg/react-18/discussions/104).

Ok, let's assume, you use `pipeToNodeWritable`. This method returns a `abort` function:

```jsx
const SERVER_RENDER_ABORT_TIMEOUT = 10000;

const { pipe, abort } = pipeToNodeWritable(
  <App />,
  onCompleteAll: () => {
    // ...
  }
);

// This is the most important part
setTimeout(() => {
    abort();
}, SERVER_RENDER_ABORT_TIMEOUT);
```

It's taken from  [Dan Abramov's demo](https://codesandbox.io/s/kind-sammet-j56ro?file=/server/render.js).

**Comment by blittle** [comment]
@artem-malko ahh, the misunderstanding I had was that I didn't realize calling abort would mean `onCompleteAll` would get called. That makes sense though. So in the case that the timeout has occurred, and I'd like to show the user with a meaningful error page, instead of just immediately calling abort, I'd send whatever HTML content I want to the user.

**Comment by blittle** [comment]
@artem-malko I don't know if the warning message from react is worth following up on?

> Warning: There was still abortable task at the root when we closed. This is a bug in React.

**Comment by gaearon** [comment]
Yeah let's keep it open to figure out what to do with the warning. @sebmarkbage?

**Comment by sebmarkbage** [comment]
That seems like a bug. The reentrancy makes it tricky to get right in the code so it's not too surprising.

A reduced test case would be helpful.

The bigger picture issue here is that if something errors (or gets aborted) before the root is flushed, we don't have a way to recover within pure React right now. It should instead trigger an error boundary and only if the error boundary suspends/fails should it close the whole stream.

**Comment by artem-malko** [comment]
@blittle You're in good hands right now)

**Comment by blittle** [comment]
@sebmarkbage I'll try and put together a simplified test case

**Comment by blittle** [comment]
My initial plan for "recovery" or at least showing an error page was to do it entirely outside react on "abort". But if we could remain within react world and fail at an error boundary, that would be ideal.

**Comment by blittle** [comment]
I have not been able to recreate the issue with the example [codesandbox that you all published](https://codesandbox.io/s/kind-sammet-j56ro?file=/server/delays.js). There must be something more specific to hydrogen where we are seeing the error. It is reproducible on [stackblitz with hydrogen](https://stackblitz.com/edit/hydrogen-v3w1xm?file=src%2Fpages%2FIndex.server.jsx). I'll keep seeing if I can narrow down what specifically is different on our end.

**Comment by yasheshcygbit** [comment]
Hi @blittle, first off all, thanks a lot for the amazing work that you guys are doing on Hydrogen, really appreciate and curious about what is coming next! Big thumbs up from a react developer who would love to see the framework used so extensively for Shopify!

Our development team has started working on Hydrogen for one of our Shopify Storefront. So far we were able to develop almost all pages without any issues except this one warning that we also got while we were trying to open Product Detail page. Can you help me know status of your issue as well and whether we will be able to use any workarounds in this case? 

Our server shows this particular warning: "Warning: There was still abortable task at the root when we closed. This is a bug in React." and our app freezes(starts showing suspense).

**Comment by blittle** [comment]
Hi @yasheshcygbit, could you open an issue on https://github.com/Shopify/hydrogen or message me on https://discord.com/invite/shopifydevs
